### ISSUE TITLE
[gfx950][MLA] Split-KV long-prefill: intuition, softmax math, launch policy, and validation gaps

### ISSUE THREAD

**Issue by raikonenfnu** [raikonenfnu/tokenspeed#74]
## Scope and status

This issue explains the optimization implemented by [`perf/gfx950-mla-prefill-split-kv`](https://github.com/raikonenfnu/tokenspeed/tree/perf/gfx950-mla-prefill-split-kv) at [`0d365a09`](https://github.com/raikonenfnu/tokenspeed/commit/0d365a09344ce102e89c2c3d7fbeb14b570b51e2).

The branch substantially improves long-context Kimi-K3 TTFT on MI355X/gfx950, but it is still experimental. Current serving validation has observed first-token divergence when multi-request prefill batches form. The committed correctness test covers batch size 1 only, so ragged `B > 1` correctness is a blocking validation gap. The performance results below should therefore be read as evidence that the parallelization strategy is promising, not as evidence that this exact implementation is ready to merge.

Related benchmark and reproduction context: #71 and #73.

## Intuition: turn one long job into several medium jobs

MLA prefill computes attention for query blocks against an increasingly long KV history. The original gfx950 kernel assigns each query block/head to a persistent CTA. That CTA walks the visible KV range serially:

```text
one query CTA:  KV tile 0 -> tile 1 -> ... -> tile T-1
```

For a long history and a relatively small chunked-prefill query grid, two things happen:

1. There may not be enough independent query CTAs to keep all 256 CUs busy for multiple waves.
2. Every CTA has a long serial critical path proportional to the number of KV tiles.

The split-KV path partitions the KV walk into `S` page-aligned ranges:

```text
query block/head
  |-- split 0: KV tiles [0, T/S)
  |-- split 1: KV tiles [T/S, 2T/S)
  |-- ...
  `-- split S-1

partial softmax summaries -> one lightweight reduction -> final output
```

This increases available CTA parallelism by approximately `S` and shortens each attention CTA's KV walk by approximately `1/S`. The tradeoff is a second kernel launch, temporary FP32 buffers, and a reduction over the split dimension.

## Attention and online-softmax math

For one query row, let the scaled attention logits be

```math
x_j = \frac{q \cdot k_j}{\sqrt{d}},
\qquad
O = \frac{\sum_j e^{x_j} v_j}{\sum_j e^{x_j}}.
```

The Gluon kernel uses base-2 exponentiation, so it internally scales logits by `1 / ln(2)`. Define `z_j = x_j / ln(2)`, making `2^{z_j} = e^{x_j}`.

For each KV partition `s`, the split CTA computes the usual numerically stable online-softmax state:

```math
m_s = \max_{j \in s} z_j,
```

```math
\ell_s = \sum_{j \in s} 2^{z_j - m_s},
```

```math
a_s = \sum_{j \in s} 2^{z_j - m_s} v_j.
```

It stores a locally normalized output and a base-2 log-normalizer:

```math
o_s = \frac{a_s}{\ell_s},
\qquad
\lambda_s = m_s + \log_2(\ell_s).
```

The reduction kernel combines the partitions without reconstructing the original logits. Let

```math
m = \max_s \lambda_s,
\qquad
\beta_s = 2^{\lambda_s - m}.
```

Then

```math
O = \frac{\sum_s \beta_s o_s}{\sum_s \beta_s}.
```

The optional natural-log LSE output is

```math
\mathrm{LSE}(x)
= \left(m + \log_2\sum_s \beta_s\right)\ln(2).
```

Why this is correct:

```math
\beta_s o_s
= 2^{m_s + \log_2 \ell_s - m}\frac{a_s}{\ell_s}
= 2^{-m}2^{m_s}a_s
= 2^{-m}\sum_{j \in s}2^{z_j}v_j.
```

The common `2^{-m}` factor cancels between numerator and denominator, yielding the original global softmax. This is mathematically equivalent to the unsplit kernel. It is not bitwise equivalent in floating point because the partitioning changes reduction order and rounding.

## Launch policy

The branch activates split-KV when

```text
max_seqlen_kv >= 8192
```

and otherwise retains the existing allocation-free persistent kernel.

For the long path:

- `BLOCK_M = 128`
- `BLOCK_N` changes from 64 to 32
- MI355X is modeled as 256 CUs
- target occupancy is four machine waves, or 1024 CTAs
- split count is capped at four

The unsplit query-grid estimate is

```math
G_0 = \left\lceil\frac{L_{q,\max}}{128}\right\rceil B H,
```

where `B` is batch size and `H` is the local query-head count. The policy computes

```math
S_{required} = \left\lceil\frac{256 \cdot 4}{G_0}\right\rceil,
```

rounds it up to a power of two, then caps it by four and by the number of available KV tiles. The current branch also forces at least two splits on the long path, even when the query grid already supplies enough CTAs.

The resulting split launch grid is

```text
(ceil(max_q / BLOCK_M) * S, batch_size, num_heads)
```

and the reduction grid is

```text
(total_query_tokens, num_heads)
```

Power-of-two split counts keep partitions balanced and make the reduction dimension tiny (`S` is normally 2 or 4).

## Why `BLOCK_N = 32` can help gfx950

The K, RoPE-K, and V tiles are double-buffered in LDS. Ignoring layout padding, their storage is proportional to

```math
2 \cdot BLOCK_N \cdot (D_K + D_{RoPE} + D_V) \cdot \text{element size}.
```

For Kimi-K3 MLA, `(D_K, D_RoPE, D_V) = (128, 64, 128)`. Reducing `BLOCK_N` from 64 to 32 therefore roughly halves this tile storage per CTA. Lower LDS pressure can permit more resident CTAs per CU, which is valuable after split-KV creates the extra CTA supply.

The cost is twice as many inner-loop KV tiles and more loop/control overhead. `BLOCK_N = 32` is therefore not intrinsically better; it is part of the occupancy/parallelism tradeoff and must be measured separately from splitting.

## Work retained from the persistent kernel

Each split still uses the optimized gfx950 building blocks already present in the kernel:

- MFMA for `QK` and `PV`;
- asynchronous global-to-LDS copies;
- double-buffered K, RoPE-K, and V tiles;
- overlap of the next tile load with current-tile compute;
- unmasked processing for fully visible causal tiles;
- masking only for the causal diagonal and final KV tail;
- base-2 online softmax for stable accumulation.

The split boundaries are KV-tile/page aligned, avoiding overlap or gaps between partitions.

## Explicit costs

For `T_q` total query tokens, `H` local heads, value dimension `D_v`, and `S` splits, the temporary workspace is

```math
4 T_q H S (D_v + 1)\ \text{bytes},
```

because both `mid_o[T_q,H,S,D_v]` and `mid_lse[T_q,H,S]` are FP32.

The long path additionally pays for:

1. two temporary allocations per attention invocation;
2. a larger CTA grid;
3. an extra reduction-kernel launch;
4. global writes and reads of partial outputs;
5. the forced `BLOCK_N = 32` inner-loop shape.

These costs occur per attention layer. A standalone attention speedup does not guarantee a whole-model or TTFT speedup, so CPU allocation/launch time and GPU kernel time must be measured separately.

## Observed end-to-end result

The clean 50K/500 non-speculative Kimi-K3 reproduction in #73 used TP8/EP8 on 8x MI355X with a 4K prefill budget. Measured split-KV results included:

| Concurrency | Mean TTFT | Output throughput |
|---:|---:|---:|
| 8 | 19.82 s | 103.19 tok/s |
| 16 | 34.06 s | 114.45 tok/s |

Relative to the older TokenSpeed row that also differed in prefill budget (8K rather than 4K), TTFT was 50.2% lower at C8 and 54.6% lower at C16. This demonstrates strong end-to-end potential, but it is not a clean kernel-only A/B because the prefill budget changed.

## Correctness gap found by serving validation

The branch's committed numerical test exercises one sequence with `q_len=128`, `kv_len=8192`. It validates the split result against an FP32 reference, including LSE, but does not exercise:

- batch size greater than one;
- ragged `cu_seqlens_q` / `cu_seqlens_kv`;
- batches formed from multiple chunked-prefill requests;
- first-token logits at whole-model level.

In controlled serving runs, identical 50K first-turn prompts were token-identical at C1, while multi-request runs produced first-token divergence for most requests. That makes later agentic trajectories and throughput non-comparable. Until the batched/ragged failure is reproduced and fixed at kernel level, the branch should not be promoted on performance alone.

### Correctness first

- [ ] Add equal-length and ragged `B=2/4` split-KV tests against an FP32 reference.
- [ ] Exercise mixed query lengths, KV lengths, prefix lengths, tail tiles, and strided output.
- [ ] Compare split and persistent whole-model first-token logits on identical prompts.
- [ ] Record top-1/top-2 IDs, logits, and margin at the first divergence.
- [ ] Require finite partial LSE/output values and verify every valid row/split is written.

### Isolate each performance lever

Measure these as independent variants:

1. persistent kernel with `BLOCK_N=64` (original control);
2. persistent kernel with `BLOCK_N=32`;
3. split with `S=2`;
4. split with `S=4`;
5. allow `S=1` when the existing query grid already provides enough CTAs;
6. target one, two, and four machine waves.

For each variant, capture:

- CPU allocation and launch time;
- split attention GPU time;
- reduction GPU time;
- per-layer MLA time;
- whole-model prefill-forward time;
- number and size of scheduled prefill chunks;
- time between chunks;
- first decode and first-token emission time.

### Promotion criteria

A final launch policy should simultaneously:

- pass batched/ragged numerical tests;
- preserve identical-prompt first-token behavior within an explained logit margin;
- improve identical-prompt `max_tokens=1` TTFT;
- improve or preserve `max_tokens=500` throughput;
- retain the 50K/500 C1-C16 gain;
- improve the frozen multi-turn agentic workload without relying on divergent trajectories.
